wallet/test/Migrations.hs

```diff
@@ -0,0 +1,206 @@
+deriving instance Eq AccountInfo_v0
+deriving instance Eq AddressInfo_v0
+deriving instance Eq WalletStorage_v2
+
+deriving instance Show WalletTip
+deriving instance Show WalletInfo
+deriving instance Show AccountInfo_v0
+deriving instance Show AddressInfo_v0
+deriving instance Show WalletStorage_v2
+
+deriving instance Arbitrary CHash
+deriving instance Arbitrary (CId Wal)
+deriving instance Arbitrary CTxId
+deriving instance Arbitrary CCoin
+
+instance Arbitrary CProfile where
+  arbitrary = CProfile <$> arbitrary
+
+instance Arbitrary CTxMeta where
+  arbitrary = CTxMeta <$> arbitrary
+
+instance Arbitrary CUpdateInfo where
+  arbitrary = CUpdateInfo
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+
+instance Arbitrary (CId Addr) where
+  arbitrary = addressToCId <$> arbitrary
+
+instance Arbitrary CAccountMeta where
+  arbitrary = CAccountMeta <$> arbitrary
+
+instance Arbitrary CWalletAssurance where
+  arbitrary = oneof
+    [ pure CWAStrict
+    , pure CWANormal
+    ]
+
+instance Arbitrary WAddressMeta where
+  arbitrary = WAddressMeta
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+
+instance Arbitrary CWAddressMeta where
+  arbitrary = CWAddressMeta
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+
+instance Arbitrary CWalletMeta where
+  arbitrary = CWalletMeta
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+
+instance Arbitrary AccountId where
+  arbitrary = AccountId
+    <$> arbitrary
+    <*> arbitrary
+
+instance Arbitrary WalletTip where
+  arbitrary = oneof
+    [ pure NotSynced
+    , SyncedWith <$> arbitrary
+    ]
+
+instance Arbitrary AddressInfo where
+  arbitrary = AddressInfo
+    <$> arbitrary
+    <*> arbitrary
+instance Arbitrary AddressInfo_v0 where
+  arbitrary = AddressInfo_v0
+    <$> arbitrary
+    <*> arbitrary
+
+instance Arbitrary AccountInfo where
+  arbitrary = AccountInfo
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+
+instance Arbitrary AccountInfo_v0 where
+  arbitrary = AccountInfo_v0
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+
+instance Arbitrary WalletInfo where
+  arbitrary = WalletInfo
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> pure HM.empty
+    <*> arbitrary
+
+instance Arbitrary WalletStorage_v2 where
+  arbitrary = WalletStorage_v2
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> pure HM.empty
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+
+tests :: TestTree
+tests = testGroup "Migration"
+    [ testProperty "migration to latest version can be reversed" $ prop_backMigrate
+    ]
+  where
+    prop_backMigrate :: WalletStorage_v2 -> Bool
+    prop_backMigrate ws = let
+        WalletStorage_Back_v2 ws' = migrate . migrate $ ws
```

Comment:
This bit is confusing at first without having called out what we're doing. We set up the instances so that the second `migrate` is going from current -> v2, so that overall it's a round trip v2 -> current and current -> v2.